attach/cid.c

```diff
@@ -0,0 +1,235 @@
+ * mapping is added to cid_map_list.
+ */
+static void cid_save_attachment(struct Body *body, struct CidMapList *cid_map_list)
+{
+  if (!body || !cid_map_list)
+    return;
+
+  char *id = mutt_param_get(&body->parameter, "content-id");
+  struct Buffer *tmpfile = mutt_buffer_pool_get();
+  struct Buffer *cid = mutt_buffer_pool_get();
+  bool has_tempfile = false;
+  FILE *fp = NULL;
+
+  if (id)
+  {
+    mutt_debug(LL_DEBUG2, "attachment found with \"Content-ID: %s\"\n", id);
+    /* get filename */
+    char *fname = mutt_str_dup(body->filename);
+    if (body->aptr)
+      fp = body->aptr->fp;
+    mutt_file_sanitize_filename(fname, fp ? true : false);
+    mailcap_expand_filename("%s", fname, tmpfile);
+    FREE(&fname);
+
+    /* save attachment */
+    if (mutt_save_attachment(fp, body, mutt_buffer_string(tmpfile), 0, NULL) == -1)
+      goto bail;
+    has_tempfile = true;
+    mutt_debug(LL_DEBUG2, "attachment with \"Content-ID: %s\" saved to file \"%s\"\n",
+               id, mutt_buffer_string(tmpfile));
+
+    /* add Content-ID to filename mapping to list */
+    mutt_buffer_printf(cid, "cid:%s", id);
+    struct CidMap *cid_map = cid_map_new(mutt_buffer_string(cid),
+                                         mutt_buffer_string(tmpfile));
+    STAILQ_INSERT_TAIL(cid_map_list, cid_map, entries);
+  }
+
+bail:
+
+  if ((fp && !mutt_buffer_is_empty(tmpfile)) || has_tempfile)
+    mutt_add_temp_attachment(mutt_buffer_string(tmpfile));
+  mutt_buffer_pool_release(&tmpfile);
+  mutt_buffer_pool_release(&cid);
+}
+
+/**
+ * cid_save_attachments - Save all attachments in a "multipart/related" group with a Content-ID
+ * @param[in]  body         First body in "multipart/related" group
+ * @param[out] cid_map_list List of Content-ID to filename mappings
+ */
+void cid_save_attachments(struct Body *body, struct CidMapList *cid_map_list)
+{
+  if (!body || !cid_map_list)
+    return;
+
+  for (struct Body *b = body; b; b = b->next)
+  {
+    if (b->parts)
+      cid_save_attachments(b->parts, cid_map_list);
+    else
+      cid_save_attachment(b, cid_map_list);
+  }
+}
+
+/**
+ * cid_to_filename - Replace Content-IDs with filenames
+ * @param filename     Path to file to replace Content-IDs with filenames
+ * @param cid_map_list List of Content-ID to filename mappings
+ */
+void cid_to_filename(struct Buffer *filename, struct CidMapList *cid_map_list)
+{
+  if (!filename || !cid_map_list)
+    return;
+
+  FILE *fp_out = NULL;
+  char *pbuf = NULL;
+  char *searchbuf = NULL;
+  char *buf = NULL;
+  char *cid = NULL;
+  size_t blen = 0;
+  struct CidMap *cid_map = NULL;
+
+  struct Buffer *tmpfile = mutt_buffer_pool_get();
+  struct Buffer *tmpbuf = mutt_buffer_pool_get();
+
+  FILE *fp_in = mutt_file_fopen(mutt_buffer_string(filename), "r");
+  if (!fp_in)
+    goto bail;
+
+  const char *suffix = mutt_strn_rfind(mutt_buffer_string(filename),
+                                       mutt_buffer_len(filename), ".");
+  if (++suffix)
```

Comment:
I think you're missing a `*`
`mutt_strn_rfind()` returns a pointer, or `NULL`.
If you prefix-increment the pointer, it will be moved on one place.
If you increment `NULL`, you'll get 0x0001, which is `true`, but it'll crash.

Reply:
Ugh! That is a bad error.

What I was intending was to ensure that filename didn't end in a `.` (i.e., next character is `\0`).

It never went wrong because in theory `filename` should always have an extension given to it already.

I think this should be safe…

```C
  if (suffix && *(suffix++))
```

Reply:
yeah, that looks a bit better